src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -211,6 +211,59 @@ protected override ModuleFieldLayout CreateValueFromKey(EcmaModule module)
                     typeOffsets: typeOffsets);
             }
 
+            private void GetElementTypeInfoGeneric(
+                EcmaModule module,
+                FieldDesc fieldDesc,
+                EntityHandle valueTypeHandle,
+                bool moduleLayout,
+                ref int alignment,
+                ref int size,
+                ref bool isGcPointerField,
+                ref bool isGcBoxedField)
+            {
+                TypeDesc fieldType = fieldDesc.FieldType;
+
+                if (fieldType.IsPrimitive)
+                {
+                    size = fieldType.GetElementSize().AsInt;
+                    alignment = size;
+                }
+                else if (fieldType.IsValueType)
+                {
+                    if (IsTypeByRefLike(valueTypeHandle, module.MetadataReader))
+                    {
+                        ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadGeneral, fieldDesc.OwningType);
+                    }
+                    if (moduleLayout && fieldType.GetTypeDefinition() is EcmaType ecmaType && ecmaType.EcmaModule != module)
+                    {
+                        // Allocate pessimistic non-GC area for cross-module fields as that's what CoreCLR does
+                        // <a href="https://github.com/dotnet/coreclr/blob/659af58047a949ed50d11101708538d2e87f2568/src/vm/ceeload.cpp#L2124">here</a>
+                        alignment = TargetDetails.MaximumPrimitiveSize;
+                        size = TargetDetails.MaximumPrimitiveSize;
+                        isGcBoxedField = true;
+                    }
+                    else if (fieldType.IsEnum)
+                    {
+                        size = fieldType.UnderlyingType.GetElementSize().AsInt;
+                        alignment = size;
+                        isGcBoxedField = false;
+                    }
+                    else
+                    {
+                        // All struct statics are boxed in CoreCLR
+                        isGcBoxedField = true;
+                    }
+                }
+                else if (fieldType.IsByRef || fieldType.IsByRefLike || fieldType.IsByReferenceOfT)
+                {
+                    ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadGeneral, fieldDesc.OwningType);
+                }
+                else
```

Comment:
Might want to also handle fieldType.IsPointer and fieldType.IsFunctionPointer for parity with ELEMENT_TYPE_FNPTR and ELEMENT_TYPE_PTR.